{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:100% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##########################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [1] Functions & Packages for annotating SNPs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##########################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcf\n",
    "\n",
    "%matplotlib inline\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import matplotlib.ticker as ticker\n",
    "from pylab import plot, show, savefig, xlim, figure, hold, ylim, legend, boxplot, setp, axes\n",
    "from itertools import compress\n",
    "from pylab import MaxNLocator\n",
    "import seaborn as sns; sns.set()\n",
    "from matplotlib.colors import LogNorm\n",
    "from matplotlib import gridspec\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import ast\n",
    "import itertools\n",
    "import seaborn as sns\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "import fastcluster\n",
    "from sklearn import cluster, datasets\n",
    "import scipy.cluster.hierarchy as hier\n",
    "from sklearn.cluster import KMeans\n",
    "import time\n",
    "import sys\n",
    "\n",
    "import Bio\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio import pairwise2\n",
    "from Bio import SeqIO\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "from Bio.SeqUtils import GC\n",
    "\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from StringIO import StringIO\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "from Bio.Seq import MutableSeq\n",
    "import itertools\n",
    "\n",
    "import networkx as nx\n",
    "import scipy\n",
    "import pickle\n",
    "\n",
    "#for exporting to Adobe Illustrator\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "mpl.rcParams['ps.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Decide on a threshold for Alternate Allele Frequencies to call SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "alt_AF_threshold = 0.20 #x% "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load regions to exclude from analysis per EBR score across H37Rv (dropping sites with EBR score < 0.9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/n/data1/hms/dbmi/farhat/Roger/dennis_mujuni_collaboration/variant_calls/H37Rv_sites_to_drop.pkl', 'rb') as f:\n",
    "    H37Rv_positions_to_drop = pickle.load(f)\n",
    "    \n",
    "#convert to a set (faster to query)\n",
    "H37Rv_positions_to_drop = set(H37Rv_positions_to_drop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cell to annotate SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Important Packages\n",
    "################################################################################################################################################################################################\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys\n",
    "import pickle\n",
    "\n",
    "import Bio\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio import SeqIO\n",
    "from StringIO import StringIO\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "from Bio.Seq import MutableSeq\n",
    "################################################################################################################################################################################################\n",
    "\n",
    "\n",
    "# Relevant Information for H37Rv sequence SNP functional annotation\n",
    "################################################################################################################################################################################################\n",
    "####### Collect all DNA and Amino Acid sequences corresponding to genes on H37Rv #######\n",
    "#load reference genome and reference annotation\n",
    "reference_genome = '/n/data1/hms/dbmi/farhat/bin/work-horse/bin/h37rv.fasta'\n",
    "for reference_genome in SeqIO.parse(reference_genome, \"fasta\"):\n",
    "    reference_genome.seq.alphabet = IUPAC.unambiguous_dna\n",
    "\n",
    "reference_genome_annotation = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/H37Rv/h37rv_genome_summary.txt', '\\t').set_index('name')\n",
    "\n",
    "####### Function to translate coding DNA sequences ####### \n",
    "def translate(gene_id, sequence):\n",
    "\n",
    "    #find which strand the gene is located on and translate\n",
    "    strand = reference_genome_annotation.loc[gene_id, 'strand']\n",
    "    if strand == '+':\n",
    "        protein_sequence = sequence.translate(table=\"Bacterial\", cds=False)\n",
    "    elif strand == '-':\n",
    "        protein_sequence = sequence.reverse_complement().translate(table=\"Bacterial\", cds=False)\n",
    "\n",
    "    return protein_sequence\n",
    "\n",
    "####### Load in dictionaries for SNP annotation #######\n",
    "with open('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/pickled_files/dicts_for_SNP_annotation/H37Rv_gene_seq_records.pickle', 'rb') as handle:\n",
    "    ref_gene_sequences_records = pickle.load(handle)\n",
    "    \n",
    "with open('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/pickled_files/dicts_for_SNP_annotation/H37Rv_protein_seq_records.pickle', 'rb') as handle:\n",
    "    ref_protein_sequences_records = pickle.load(handle)\n",
    "    \n",
    "with open('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/pickled_files/dicts_for_SNP_annotation/H37Rv_coord_gene_mapping.pickle', 'rb') as handle:\n",
    "    ReferencePosition_Gene_mapping = pickle.load(handle)\n",
    "    \n",
    "####### get Gene Categories #######\n",
    "gene_categories = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/gene_categories/gene_categories.csv').set_index('name')\n",
    "gene_categories_dict = dict([gene_id , gene_category] for gene_id, gene_category in zip(list(gene_categories.gene_id) , list(gene_categories.Gene_Category)))\n",
    "\n",
    "####### get Gene Symbols #######\n",
    "gene_symbol_dict = dict([gene_id , gene_symbol] for gene_id, gene_symbol in zip(list(reference_genome_annotation.symbol.index) , list( reference_genome_annotation.symbol )))\n",
    "################################################################################################################################################################################################\n",
    "\n",
    "\n",
    "# Function to annotate Intergenic SNPs\n",
    "################################################################################################################################################################################################\n",
    "def find_flanking_genes_for_intergenic_region(intergenic_ref_pos): \n",
    "\n",
    "    #this function finds the genes flagging an intergenic region given a reference position\n",
    "\n",
    "    #find gene immediately in the 5' direction\n",
    "    for i in range(0 , 100000):\n",
    "\n",
    "        #move toward 5' direction\n",
    "        if ReferencePosition_Gene_mapping[intergenic_ref_pos - i] != []:\n",
    "\n",
    "            gene_to_left = ReferencePosition_Gene_mapping[intergenic_ref_pos - i][0]\n",
    "            break\n",
    "\n",
    "    #find gene immediately in the 3' direction       \n",
    "    for i in range(0 , 100000):\n",
    "\n",
    "        #move toward 3' direction\n",
    "        try:\n",
    "            if ReferencePosition_Gene_mapping[intergenic_ref_pos + i] != []:\n",
    "\n",
    "                gene_to_right = ReferencePosition_Gene_mapping[intergenic_ref_pos + i][0]\n",
    "                break\n",
    "        \n",
    "        #KeyError means we have hit the 'end' of the chromosome, the intergenic region at then end of H37Rv in 5' > 3' orientation \n",
    "        #since TB chromosome is circular the gene to the 'right' is Rv0001    \n",
    "        except KeyError:\n",
    "            \n",
    "            gene_to_right = 'Rv0001'\n",
    "            break\n",
    "    \n",
    "    return gene_to_left + '_' + gene_to_right\n",
    "################################################################################################################################################################################################\n",
    "\n",
    "\n",
    "# Function to determine whether SNPs are Synonymous or Non-Synonymous; Returns gene coordinate, codon position, AA changes, Gene Category & Symbol\n",
    "################################################################################################################################################################################################\n",
    "def SNP_annotate(ref_seq_position , alt_allele_i):\n",
    "    \n",
    "    '''\n",
    "    This function takes as input a reference position on H37Rv located within a \n",
    "    gene and an alternate allele and returns whether the base change \n",
    "    would correspond to a different Amino Acid sequence that results \n",
    "    from translating the DNA sequence into an AA sequence.\n",
    "    \n",
    "    '''\n",
    "    gene_intergenic_id_list = []\n",
    "    genomic_coord_list = []\n",
    "    gene_category_list = []\n",
    "    gene_symbol_list = []\n",
    "    Syn_NSyn_list = []\n",
    "    AA_change_list = []\n",
    "    \n",
    "    #get the Reference Allele from the complete H37Rv reference genome, indexing starts from 0\n",
    "    ref_allele_i = reference_genome.seq[int(ref_seq_position) - 1] \n",
    "    \n",
    "    #find the gene that SNP occurs on; check list corresponding to H37Rv coordinate to see if there are any genes associated with RefPosition\n",
    "    if len(ReferencePosition_Gene_mapping[ref_seq_position]) > 0:\n",
    "\n",
    "        #iterate through all genes that ReferencePosition is mapped to (i.e. SNP might correspond to 2 genes)\n",
    "        for gene_intergenic_id in ReferencePosition_Gene_mapping[ref_seq_position]:\n",
    "\n",
    "            #find genomic coordinate of SNP relative to gene (subtract 1 since reference seq starts counting at 1)\n",
    "            gene_relative_coord = (ref_seq_position - 1) - min( reference_genome_annotation.loc[gene_intergenic_id , 'chromStart'] , reference_genome_annotation.loc[gene_intergenic_id , 'chromEnd'] )\n",
    "            \n",
    "            #find the genomic coordinate (relative to the gene, in the 5' to 3' direction)\n",
    "            strand = reference_genome_annotation.loc[gene_intergenic_id, 'strand']\n",
    "            if strand == '+':\n",
    "                 genomic_5_to_3_coord = (ref_seq_position) - reference_genome_annotation.loc[gene_intergenic_id , 'chromStart']\n",
    "\n",
    "            elif strand == '-':\n",
    "                 genomic_5_to_3_coord = (reference_genome_annotation.loc[gene_intergenic_id , 'chromEnd']) - (ref_seq_position-1)\n",
    "                    \n",
    "            #find gene category (if one exists)\n",
    "            try:\n",
    "                gene_category_i = gene_categories_dict[gene_intergenic_id]\n",
    "            except KeyError:\n",
    "                gene_category_i = 'None'\n",
    "            \n",
    "            #find gene symbol (if one exists)\n",
    "            try:\n",
    "                gene_symbol_i = gene_symbol_dict[gene_intergenic_id]\n",
    "            except KeyError:\n",
    "                gene_symbol_i = 'None'\n",
    "            \n",
    "            #alternate allele is an actual base\n",
    "            if alt_allele_i in ['A','C','G','T']:\n",
    "\n",
    "                #translate into protein sequence with the SNP in place if not InDel or intergenic region\n",
    "                SNP_change = alt_allele_i\n",
    "\n",
    "                #ALTERNATE allele (is it Syn or NSyn?)\n",
    "                #get sequence from dictionary of sequences (and convert to mutable object)\n",
    "                test_gene_sequence = ref_gene_sequences_records[gene_intergenic_id].seq.tomutable()\n",
    "\n",
    "                #change reference gene sequence by the SNP in the query sequence\n",
    "                test_gene_sequence[int(gene_relative_coord)] = SNP_change\n",
    "\n",
    "                #convert back immutable object\n",
    "                test_gene_sequence = test_gene_sequence.toseq()\n",
    "\n",
    "                #translate sequence into amino acid seq\n",
    "                test_protein_sequence = translate(gene_intergenic_id , test_gene_sequence)\n",
    "\n",
    "                #store the H37Rv AA seq to compare against\n",
    "                H37Rv_AA_sequence = ref_protein_sequences_records[gene_intergenic_id].seq\n",
    "\n",
    "                #get the codon number where the SNP occurs within\n",
    "                ## take the genomic coordinate (relative to the gene, in the 5' to 3' direction), divide by 3, then take the ceiling of this number (will be fraction if SNP occurs in 1st or 2nd position on codon)\n",
    "                strand = reference_genome_annotation.loc[gene_intergenic_id, 'strand']\n",
    "                if strand == '+':\n",
    "                     genomic_5_to_3_coord = (ref_seq_position) - reference_genome_annotation.loc[gene_intergenic_id , 'chromStart']\n",
    "\n",
    "                elif strand == '-':\n",
    "                     genomic_5_to_3_coord = (reference_genome_annotation.loc[gene_intergenic_id , 'chromEnd']) - (ref_seq_position-1)\n",
    "\n",
    "                codon_coord = int(np.ceil( float( genomic_5_to_3_coord) / 3.0 ))\n",
    "\n",
    "                #compare to AA seq of original gene\n",
    "                if test_protein_sequence == H37Rv_AA_sequence:\n",
    "\n",
    "                    SNP_type = 'S'\n",
    "\n",
    "                    #get the AA before & after\n",
    "                    AA_change = H37Rv_AA_sequence[codon_coord-1] + str(codon_coord) + test_protein_sequence[codon_coord-1]\n",
    "\n",
    "                else:\n",
    "                    SNP_type = 'N'\n",
    "\n",
    "                    #get the AA before & after\n",
    "                    AA_change = H37Rv_AA_sequence[codon_coord-1] + str(codon_coord) + test_protein_sequence[codon_coord-1]\n",
    "                    \n",
    "            #alternate allele is a dummy (Base Call completely supports the Reference Allele)       \n",
    "            else:\n",
    "                \n",
    "                SNP_type = 'None'\n",
    "                AA_change = 'None'\n",
    "\n",
    "            #store relevant info in lists    \n",
    "            gene_intergenic_id_list.append(gene_intergenic_id)\n",
    "            genomic_coord_list.append(genomic_5_to_3_coord)\n",
    "            gene_category_list.append(gene_category_i)\n",
    "            gene_symbol_list.append(gene_symbol_i)\n",
    "            Syn_NSyn_list.append(SNP_type)\n",
    "            AA_change_list.append(AA_change)\n",
    "    \n",
    "    #if no gene in H37Rv corresponds to the Reference Position for SNP, then SNP must be intergenic\n",
    "    else:\n",
    "        \n",
    "        gene_intergenic_id = find_flanking_genes_for_intergenic_region(ref_seq_position)\n",
    "        genomic_5_to_3_coord = 'None'\n",
    "        gene_category_i = 'None'\n",
    "        gene_symbol_i = 'None'\n",
    "        SNP_type = 'I'\n",
    "        AA_change = 'None'\n",
    "        \n",
    "        #store relevant info in lists    \n",
    "        gene_intergenic_id_list.append(gene_intergenic_id)\n",
    "        genomic_coord_list.append(genomic_5_to_3_coord)\n",
    "        gene_category_list.append(gene_category_i)\n",
    "        gene_symbol_list.append(gene_symbol_i)\n",
    "        Syn_NSyn_list.append(SNP_type)\n",
    "        AA_change_list.append(AA_change)\n",
    "    \n",
    "    #if there is only a single gene associated with this SNP, just return the individual elememts\n",
    "    if len(gene_intergenic_id_list) == 1:\n",
    "        return [ref_allele_i , gene_intergenic_id , genomic_5_to_3_coord , gene_category_i , gene_symbol_i , SNP_type , AA_change]\n",
    "    \n",
    "    #else if there are two genes associated with this SNP, return elements for each SNP annotation in a list\n",
    "    elif len(gene_intergenic_id_list) > 1:\n",
    "        return [ref_allele_i , gene_intergenic_id_list , genomic_coord_list , gene_category_list , gene_symbol_list , Syn_NSyn_list , AA_change_list]\n",
    "################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function to get parse through and annotate SNPs (filtered for $AF$, MGE and low EBR score regions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_and_filter_SNPs(sample_tag , alt_AF_threshold):\n",
    "\n",
    "    '''\n",
    "    This function annotates SNPs (Base Calls) extracted from VCF files. \n",
    "    This function drops SNPs called with an alternate allele frequency (AF) less than x%.\n",
    "    This function also drops SNPs in regions flagged as Mobile Genetic Elements & Regions of poor Illumina mapping / variant calling\n",
    "    per Empirical Base Recall (EBR) scores across H37Rv.\n",
    "    '''\n",
    "\n",
    "    ################################################################################\n",
    "    ### get SNPs that were extracted from VCF files\n",
    "    ################################################################################\n",
    "\n",
    "    #load in the differing Base Calls for the isolate pair from pickle\n",
    "    SNPs_from_VCF = pd.read_pickle(parent_output_dir + sample_tag +  '/' + sample_tag + '_SNP_calls.pkl')\n",
    "\n",
    "    ################################################################################\n",
    "    ### Drop SNPs with change in AF < x%\n",
    "    ################################################################################\n",
    "\n",
    "    #FILTER out paired Base Calls that have an in Alternate Allele Frequency of less than x%\n",
    "    SNPs_from_VCF = SNPs_from_VCF[SNPs_from_VCF.alt_AF >= alt_AF_threshold]\n",
    "\n",
    "    #reset index of filtered SNP DataFrame\n",
    "    SNPs_from_VCF.reset_index(inplace = True, drop = True)\n",
    "\n",
    "    ################################################################################\n",
    "    ### Drop SNPs in regions with low EBR scores\n",
    "    ################################################################################\n",
    "\n",
    "    #Drop Base Calls in H37Rv sites with low EBR score (make sure there is at least 1 SNP)\n",
    "    if np.shape(SNPs_from_VCF)[0] > 0:\n",
    "\n",
    "        #create a boolean filter for SNPs to keep\n",
    "        SNPs_to_keep_filter = [SNP_i_ref_pos not in H37Rv_positions_to_drop for SNP_i_ref_pos in SNPs_from_VCF.ref_position]\n",
    "\n",
    "        #filter out SNPs in H37Rv sites with low EBR scores and reset index\n",
    "        SNPs_from_VCF = SNPs_from_VCF[SNPs_to_keep_filter]\n",
    "        SNPs_from_VCF.reset_index(inplace = True, drop = True)\n",
    "\n",
    "    ################################################################################\n",
    "    ### Annotate SNPs & Drop SNPs in MGE regions\n",
    "    ################################################################################\n",
    "\n",
    "    gene_id_list = []\n",
    "    gene_coord_list = []\n",
    "    gene_category_list = []\n",
    "    gene_symbol_list = []\n",
    "    SNP_ftype_list = []\n",
    "    AA_change_list = []\n",
    "\n",
    "    #Annotate Filtered Base Calls (make sure there is at least 1 SNP)\n",
    "    if np.shape(SNPs_from_VCF)[0] > 0:\n",
    "\n",
    "        for ref_position_i , alt_base_i in zip(list(SNPs_from_VCF.ref_position) , list(SNPs_from_VCF.alt_base)):\n",
    "\n",
    "            #annotate SNP\n",
    "            gene_id_i , gene_coord_i , gene_category_i , gene_symbol_i , SNP_ftype_i , AA_change_i = SNP_annotate(ref_position_i , alt_base_i)[1:]\n",
    "\n",
    "            gene_id_list.append(gene_id_i)\n",
    "            gene_coord_list.append(gene_coord_i)\n",
    "            gene_category_list.append(gene_category_i)\n",
    "            gene_symbol_list.append(gene_symbol_i)\n",
    "            SNP_ftype_list.append(SNP_ftype_i)\n",
    "            AA_change_list.append(AA_change_i)\n",
    "\n",
    "        #create columns to store SNP annotation info\n",
    "        SNPs_from_VCF['gene_id'] = gene_id_list\n",
    "        SNPs_from_VCF['gene_coord'] = gene_coord_list\n",
    "        SNPs_from_VCF['gene_category'] = gene_category_list\n",
    "        SNPs_from_VCF['gene_symbol'] = gene_symbol_list\n",
    "        SNPs_from_VCF['SNP_ftype'] = SNP_ftype_list\n",
    "        SNPs_from_VCF['AA_change'] = AA_change_list\n",
    "\n",
    "        #FILTER out Base Calls in MGE regions (Mobile Genentic Elements)\n",
    "        SNPs_to_drop_filter = [] #True if SNP is located within an MGE region\n",
    "\n",
    "        for gene_id_i in list(SNPs_from_VCF.gene_category):\n",
    "\n",
    "            #only 1 or 0 genes associated with this SNP\n",
    "            if (type(gene_id_i) == str) and (gene_id_i == 'Mobile Genetic Element'):\n",
    "\n",
    "                SNPs_to_drop_filter.append(True)\n",
    "\n",
    "            #two genes associated with this SNP\n",
    "            elif (type(gene_id_i) == list) and ('Mobile Genetic Element' in gene_id_i):\n",
    "\n",
    "                SNPs_to_drop_filter.append(True)\n",
    "\n",
    "            #SNP not in an MGE region so don't drop\n",
    "            else:\n",
    "\n",
    "                SNPs_to_drop_filter.append(False)\n",
    "\n",
    "        #create a boolean filter for SNPs to keep\n",
    "        SNPs_to_keep_filter = [not MGE_SNP for MGE_SNP in SNPs_to_drop_filter]\n",
    "\n",
    "        #filter out SNPs in MGE regions and reset index\n",
    "        SNPs_from_VCF = SNPs_from_VCF[SNPs_to_keep_filter]\n",
    "        SNPs_from_VCF.reset_index(inplace = True, drop = True)\n",
    "\n",
    "    #No SNPs detected between this pair of isolates (empty DataFrame)\n",
    "    else:\n",
    "\n",
    "        SNPs_from_VCF['gene_id'] = \"\"\n",
    "        SNPs_from_VCF['gene_coord'] = \"\"\n",
    "        SNPs_from_VCF['gene_category'] = \"\"\n",
    "        SNPs_from_VCF['gene_symbol'] = \"\"\n",
    "        SNPs_from_VCF['SNP_ftype'] = \"\"\n",
    "        SNPs_from_VCF['AA_change'] = \"\"\n",
    "\n",
    "    #drop unnecessary columns\n",
    "    SNPs_from_VCF.drop(['quality','SNP_type','PASS_filter','INFO'] , axis = 1 , inplace = True)\n",
    "\n",
    "    return SNPs_from_VCF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##########################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [2] Use functions above to annotate & filter SNPs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##########################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "#### Import Sample Annotation file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_annotation = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/dennis_mujuni_collaboration/annotation CSV files/sample_annotation.csv' , sep  = ',').set_index('tag')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### output where *pilon*-output is stored"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "parent_output_dir = '/n/data1/hms/dbmi/farhat/Roger/dennis_mujuni_collaboration/JankyPipe/output/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>sample_num</th>\n",
       "      <th>fastq_files</th>\n",
       "      <th>coverage</th>\n",
       "      <th>lineage_call</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S0</th>\n",
       "      <td>Undetermined</td>\n",
       "      <td>0</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/Roger/dennis_mujuni_c...</td>\n",
       "      <td>25.5521</td>\n",
       "      <td>4.6.1.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S1</th>\n",
       "      <td>TC100810</td>\n",
       "      <td>1</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/Roger/dennis_mujuni_c...</td>\n",
       "      <td>151.0760</td>\n",
       "      <td>4.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S2</th>\n",
       "      <td>TC94815</td>\n",
       "      <td>2</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/Roger/dennis_mujuni_c...</td>\n",
       "      <td>60.5541</td>\n",
       "      <td>4.3.4.2.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S3</th>\n",
       "      <td>TC101527</td>\n",
       "      <td>3</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/Roger/dennis_mujuni_c...</td>\n",
       "      <td>70.5301</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S4</th>\n",
       "      <td>LB95669</td>\n",
       "      <td>4</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/Roger/dennis_mujuni_c...</td>\n",
       "      <td>82.3331</td>\n",
       "      <td>2.2.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        sample_id  sample_num  \\\n",
       "tag                             \n",
       "S0   Undetermined           0   \n",
       "S1       TC100810           1   \n",
       "S2        TC94815           2   \n",
       "S3       TC101527           3   \n",
       "S4        LB95669           4   \n",
       "\n",
       "                                           fastq_files  coverage lineage_call  \n",
       "tag                                                                            \n",
       "S0   /n/data1/hms/dbmi/farhat/Roger/dennis_mujuni_c...   25.5521      4.6.1.2  \n",
       "S1   /n/data1/hms/dbmi/farhat/Roger/dennis_mujuni_c...  151.0760          4.9  \n",
       "S2   /n/data1/hms/dbmi/farhat/Roger/dennis_mujuni_c...   60.5541    4.3.4.2.1  \n",
       "S3   /n/data1/hms/dbmi/farhat/Roger/dennis_mujuni_c...   70.5301            3  \n",
       "S4   /n/data1/hms/dbmi/farhat/Roger/dennis_mujuni_c...   82.3331        2.2.1  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_annotation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(24, 5)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(sample_annotation)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Collect SNPs passing Difference in Alternate Allele Frequency Threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/rv76/anaconda2/lib/python2.7/site-packages/Bio/Seq.py:2715: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n"
     ]
    }
   ],
   "source": [
    "SNPs_from_samples_df_list = []\n",
    "\n",
    "sample_index = 0\n",
    "\n",
    "#iterate through isolate pairs, collect all SNP variants called in each sample\n",
    "for sample_tag in sample_annotation.index:\n",
    "    \n",
    "    #retrieve filtered & annotated SNPs\n",
    "    SNP_variants_sample_i = annotate_and_filter_SNPs(sample_tag , alt_AF_threshold)\n",
    "    \n",
    "    #store relevant SNP info in list of DataFrames\n",
    "    SNPs_from_samples_df_list.append(SNP_variants_sample_i)\n",
    "    sample_index += 1\n",
    "    \n",
    "    print sample_index\n",
    "        \n",
    "#concatenate DataFrames for each sample into 1 DataFrame\n",
    "SNPs_from_all_samples_df = pd.concat(SNPs_from_samples_df_list , axis = 0)\n",
    "\n",
    "SNPs_from_all_samples_df.reset_index(inplace = True , drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_base</th>\n",
       "      <th>alt_base</th>\n",
       "      <th>ref_position</th>\n",
       "      <th>alt_AF</th>\n",
       "      <th>depth</th>\n",
       "      <th>tag</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_coord</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>SNP_ftype</th>\n",
       "      <th>AA_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1849</td>\n",
       "      <td>0.38</td>\n",
       "      <td>59</td>\n",
       "      <td>S0</td>\n",
       "      <td>Rv0001_Rv0002</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>I</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1977</td>\n",
       "      <td>0.90</td>\n",
       "      <td>50</td>\n",
       "      <td>S0</td>\n",
       "      <td>Rv0001_Rv0002</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>I</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>7362</td>\n",
       "      <td>0.93</td>\n",
       "      <td>26</td>\n",
       "      <td>S0</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>61</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>N</td>\n",
       "      <td>E21Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>7585</td>\n",
       "      <td>0.88</td>\n",
       "      <td>19</td>\n",
       "      <td>S0</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>284</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>N</td>\n",
       "      <td>S95T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9143</td>\n",
       "      <td>0.25</td>\n",
       "      <td>24</td>\n",
       "      <td>S0</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>1842</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>S</td>\n",
       "      <td>I614I</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ref_base alt_base  ref_position  alt_AF  depth tag        gene_id  \\\n",
       "0        C        A          1849    0.38     59  S0  Rv0001_Rv0002   \n",
       "1        A        G          1977    0.90     50  S0  Rv0001_Rv0002   \n",
       "2        G        C          7362    0.93     26  S0         Rv0006   \n",
       "3        G        C          7585    0.88     19  S0         Rv0006   \n",
       "4        T        C          9143    0.25     24  S0         Rv0006   \n",
       "\n",
       "  gene_coord          gene_category gene_symbol SNP_ftype AA_change  \n",
       "0       None                   None        None         I      None  \n",
       "1       None                   None        None         I      None  \n",
       "2         61  Antibiotic Resistance        gyrA         N      E21Q  \n",
       "3        284  Antibiotic Resistance        gyrA         N      S95T  \n",
       "4       1842  Antibiotic Resistance        gyrA         S     I614I  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNPs_from_all_samples_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(24782, 12)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNPs_from_all_samples_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Save output as a CSV files and pickled DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNPs_from_all_samples_df.to_csv('/n/data1/hms/dbmi/farhat/Roger/dennis_mujuni_collaboration/variant_calls/SNPs/SNP_calls_across_all_samples.csv')\n",
    "SNPs_from_all_samples_df.to_pickle('/n/data1/hms/dbmi/farhat/Roger/dennis_mujuni_collaboration/variant_calls/SNPs/SNP_calls_across_all_samples.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
